### PR TITLE
Use case search as caseflow homepage if flag set

### PR DESCRIPTION
**Pull request by lowellrex**
Connects #5084.

Serve the case search page as the caseflow homepage for signed in users who have the `case_search_home_page` flag enabled. All other users will continue to be served the help page.

Below is a screenshot of how the page looks.
image

I've been testing by manually enabling and disabling the following toggles in the rails console:
* queue_case_search
* case_search_home_page
* queue_welcome_gate
* queue_phase_two

### CODE REVIEW

File: client/app/queue/CaseListTable.jsx [hunk 1]
```diff
@@ -27,7 +27,7 @@ class CaseListTable extends React.PureComponent {
     this.props.setActiveAppeal(appeal);
     this.props.setBreadcrumbs({
       breadcrumb: `< Back to ${appeal.attributes.veteran_full_name}'s case list`,
-      path: '/queue'
+      path: window.location.pathname
```
Comment (lowellrex) [inline client/app/queue/CaseListTable.jsx]: Need this because we can be entering a case details view through the search results page we reached by way of the queue view or the new search view on the caseflow homepage.

File: client/app/queue/CaseListTable.jsx [hunk 2]
```diff
@@ -29,7 +29,7 @@ class CaseListTable extends React.PureComponent {
     this.props.setActiveAppeal(appeal);
     this.props.setBreadcrumbs({
       breadcrumb: sprintf(COPY.BACK_TO_SEARCH_RESULTS_LINK_LABEL, appeal.attributes.veteran_full_name),
-      path: '/queue'
+      path: window.location.pathname
```
Comment (lowellrex) [inline client/app/queue/CaseListTable.jsx]: Need this because we could reach this view through the search box that is now (as of this PR) available at the `/` and `/queue` URLs.

File: client/app/queue/CaseListView.jsx [hunk 3]
```diff
@@ -21,11 +23,25 @@ const backLinkStyling = css({
...
 
+const horizontalRuleStyling = css({
+  border: 0,
+  borderTop: `1px solid ${COLORS.GREY_LIGHT}`,
+  marginTop: '5rem',
+  marginBottom: '5rem'
+});
+
 class CaseListView extends React.PureComponent {
+  shouldShowBreadcrumbs = () => this.props.caseList.receivedAppeals.length > 0 || this.props.errorType;
```
Comment (lowellrex) [inline client/app/queue/CaseListView.jsx]: Need this to hide breadcrumbs on this view when it is the case search home page at the `/` URL.

File: client/app/queue/QueueLoadingScreen.jsx [hunk 4]
```diff
@@ -73,6 +73,13 @@ class QueueLoadingScreen extends React.PureComponent {
   reload = () => window.location.reload();
 
   render = () => {
+    // We expect users with the case_search_home_page feature toggle to be able to access the queue detail view (which
+    // is wrapped by this component) even though they will not have queues. Return early to avoid making the request
+    // for queues that would happen as a result of createLoadPromise().
+    if (this.props.featureToggles.case_search_home_page) {
+      return this.props.children;
+    }
```
Comment (lowellrex) [inline client/app/queue/QueueLoadingScreen.jsx]: I don't love putting this logic down here in the `QueueLoadingScreen`, but the only alternative that I could think of was using this feature toggle to conditionally wrap the `routedQueueDetail` routine in `QueueApp` which is maybe more direct but a lot more code. I'd love to hear any other ideas on how to accomplish this!

**Comment by amprokop** [comment]
@lowellrex @laurjpeterson @sneha-pai — should we put a period since it's a full sentence?

"Enter a 9-digit Veteran ID to search for all available cases for a Veteran."

**Comment by amprokop** [comment]
actually @lowellrex 🤔 i don't see the copy in the screenshot in the PR. is the screenshot out of date?

File: app/controllers/root_controller.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
+class RootController < ApplicationController
```
Comment (amprokop) [inline app/controllers/root_controller.rb]: nit: 

hmm — `HomeController`? `CaseSearchController`? `RootController` made me think ApplicationController was inheriting from this

**Comment by lowellrex** [comment]
@amprokop That is pre-existing copy: https://github.com/department-of-veterans-affairs/caseflow/blob/d87391f65adcb64b9af01839aee5ee8cd6fbffa0/COPY.json#L2

File: app/controllers/root_controller.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  skip_before_action :verify_authentication
+
+  # rubocop:disable Metrics/CyclomaticComplexity
+  def index
+    if current_user && verify_authentication
```
[the remark this answers is not in the document]
Comment (amprokop) [reply app/controllers/root_controller.rb]: I just realized the logic of this would be slightly different — with this PR, if users currently come to the homepage who have `queue_welcome_gate` enabled but NOT `queue_case_search` enabled, they will be redirected to `/queue`.

With my proposed change, these users would be directed `/help`. That seems more correct to me (we should control the behavior of Case Search users with `queue_case_search`). Thoughts?

File: app/controllers/root_controller.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+        render("queue/index") && return
+      end
+
+      if user_can_access_queue?
+        redirect_to("/queue") && return
+      end
+    end
+
+    redirect_to("/help")
+  end
```
Comment (amprokop) [inline app/controllers/root_controller.rb]: I guess overall I think this file would be easier to reason about if we simplified it to:

```
return redirect to("/help") unless current_user && current_user.authenticated? && feature_enabled?(:case_search_home_page)
return redirect_to("/queue) if user_can_access_queue?
render("queue/index")
```

File: client/app/queue/QueueLoadingScreen.jsx (hunk 4 above)
Comment (amprokop) [reply client/app/queue/QueueLoadingScreen.jsx]: Interesting. This has a side effect, right? If we give a user the `case_search_home_page` feature toggle, then that user will no longer be able to complete tasks from the details view, correct?

File: client/app/queue/SearchBar.jsx [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    },
+    '& .cf-submit': {
+      width: '10.5rem'
+    }
+  }
+});
+
+class SearchBar extends React.PureComponent {
+  render() {
+    if (this.props.shouldUseQueueCaseSearch) {
```
Comment (amprokop) [inline client/app/queue/SearchBar.jsx]: can we remove this conditional? aren't we using queue case search for everyone now?

File: client/app/queue/QueueLoadingScreen.jsx (hunk 4 above)
Comment (amprokop) [reply client/app/queue/QueueLoadingScreen.jsx]: So it seems like the problem is:

When users come from case search to the details screen, they don't need to see task info, but the QueueLoadingScreen loads appeals and tasks for a user.

Currently the `/queue/:user_id` endpoint throws an error if the user isn't an attorney or judge.

We could fix this local problem in a few ways:

- Return `[]` instead of an error from `/queue/:user_id` if the user doesn't have queue access 
- Conditionally fire off this request from the frontend somehow, only if we need task information

It begs the question of whether we should refactor so QueueLoadingScreen doesn't wrap everything, or make QueueLoadingScreen only fetch currently needed data. Last night when we were talking, @lowellrex, I had the impression that a frontend refactor was the way to go. But after our discussion when you made some opposite points, it may be desirable to keep the frontend the way it is but ensure our `GET` endpoints from the backend don't return errors.

Comment (amprokop) [reply client/app/queue/QueueLoadingScreen.jsx]: @mdbenjam @dannysellers @evankroske might have thoughts here. Let's connect on this after standup.

Comment (lowellrex) [reply client/app/queue/QueueLoadingScreen.jsx]: Exactly

Comment (lowellrex) [reply client/app/queue/QueueLoadingScreen.jsx]: Exactly. It is my understanding that these are exclusive groups. Somebody will either have queue access (`queue_welcome_gate` feature toggle) or case search home page access (`case_search_home_page`) but not both.

File: app/controllers/root_controller.rb (hunk 7 above)
Comment (lowellrex) [reply app/controllers/root_controller.rb]: Refactored. It had to be a little more complex than the above suggestion to achieve the same behaviour as the previous version, though I guess we could change what the behaviour of this function is as well if we'd like. For this refactor I left the behaviour as is.

File: app/controllers/root_controller.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (lowellrex) [reply app/controllers/root_controller.rb]: Link to the slack discussion we had on the behaviour of this function/ the root route: https://dsva.slack.com/archives/C54QCEHAL/p1525983257000558

File: client/app/queue/SearchBar.jsx (hunk 8 above)
Comment (lowellrex) [reply client/app/queue/SearchBar.jsx]: Re: https://github.com/department-of-veterans-affairs/caseflow/pull/5470#discussion_r188091165

Correct. It is my understanding that these are exclusive groups. Somebody will either have queue access (`queue_welcome_gate` feature toggle) or case search home page access (`case_search_home_page`) but not both.

Comment (lowellrex) [reply client/app/queue/SearchBar.jsx]: Re: https://github.com/department-of-veterans-affairs/caseflow/pull/5470#discussion_r188091997

It does not look like `queue_case_search` is turned on for all users yet: https://github.com/department-of-veterans-affairs/appeals-deployment/blob/c78798cdaa53160ae926601b26439f95ab43be7a/ansible/vars/features-config.yml#L200

File: client/app/queue/QueueLoadingScreen.jsx (hunk 4 above)
Comment (lowellrex) [reply client/app/queue/QueueLoadingScreen.jsx]: I'm not at all opposed to a refactor of the frontend (I think we are overloading `QueueLoadingScreen` at the moment and find `QueueApp`'s ["routed..." render functions](https://github.com/department-of-veterans-affairs/caseflow/blob/e6db9ba06400efdc0efc3d2d3bb58eededbb4ac1/client/app/queue/QueueApp.jsx#L33) particularly hard to parse and think we may be well served to move those pups into their own components).

However, I think the most expedient solution is to return an empty array from `/queue/:user_id` when the requestor has the `case_search_home_page` feature toggle enabled.

Comment (mdbenjam) [reply client/app/queue/QueueLoadingScreen.jsx]: Why are `queue_welcome_gate` and `case_search_home_page` exclusive groups? I'm hesitant about splitting the QueueDetails page into two versions. One from when you get to it from the welcome gate, and one from when you get to it from the search page. Why do these need to be different? Can't we just show task information in both cases as long as the user has the phase_two feature toggle set?

Comment (lowellrex) [reply client/app/queue/QueueLoadingScreen.jsx]: I actually guess they don't have to be exclusive groups, we could instead have `queue_welcome_gate` take precedence over `case_search_home_page` when deciding whether to redirect to `/queue` when somebody navigates to the root route or whether to remain at that root route and show the case search home page. Similarly, we could use the `queue_welcome_gate` feature flag to determine whether to send the request to `/queue/:user_id`. I like that approach.

Comment (mdbenjam) [reply client/app/queue/QueueLoadingScreen.jsx]: When you say navigating to the root route, do you mean clicking on the caseflow logo in the top left? Or something else? What is the final state supposed to be when these flags are on for all users? Are people going to be directed to `/queue` or `/`.

File: client/app/queue/SearchBar.jsx (hunk 8 above)
Comment (amprokop) [reply client/app/queue/SearchBar.jsx]: good catch. cool, let's see if we can turn it on. we can remove the conditional in a later PR if you prefer

File: app/controllers/application_controller.rb [hunk 9]
```diff
@@ -87,12 +87,18 @@ def certification_header(title)
   end
   helper_method :certification_header
 
-  def verify_queue_access
+  def current_user_can_access_queue?
```
Comment (amprokop) [inline app/controllers/application_controller.rb]: how about just `can_access_queue?` current user is pretty implicit here

File: client/app/queue/QueueLoadingScreen.jsx [hunk 10]
```diff
@@ -73,6 +73,12 @@ class QueueLoadingScreen extends React.PureComponent {
   reload = () => window.location.reload();
 
   render = () => {
+    // If the current user cannot access queue return early to avoid making the request for queues that would happen
```
Comment (amprokop) [inline client/app/queue/QueueLoadingScreen.jsx]: hmm this still feels like a hack to me. 

defer to @mdbenjam on this, but maybe we should solve this with a change the backend route — instead of `/queue/:user_id` returning an error, it should return an empty array for non-attorney/judge users.

Comment (lowellrex) [reply client/app/queue/QueueLoadingScreen.jsx]: Took a stab at making this change on the backend but it is not trivial (requires bypassing `verify_queue_access` in addition to a fair amount of other code changes targeted more directly at the behaviour we want to change). Went ahead and created a ticket to address this issue in a separate PR: https://github.com/department-of-veterans-affairs/caseflow/issues/5551
